File: metric.py
import numpy as np

class Metric:
    """
    Metric properties and calculation
    polar metric special
    """
    
    origin = (0,0)

    def __init__(self, origin):
       self.origin = origin

    def compute_metric(self, position, t="polar"):
        """
        Compute covariant metric at a point of a given type
        """
        o = np.array(self.origin)
        p = np.array(position)

        g = np.zeros((2,2))

        if t=="polar":
            g[0,0] = 1
            v = p-o
            g[1,1] = (np.sin(v[0]*np.pi)**2)

        elif t=="a1":
            g[0,0] = 1
            v = p-o
            g[1,1] = v[0]

        return g
        

File: src/metric.cpp
/*
    This file is part of 2dgrpt.

    2dgrpt is my program to test some fast geodesic algorithms.

    Copyright (c) 2018 by Kumar Ayush
*/

#include "metric.hpp"

using namespace grpt;

//TODO: Overload () operator to do this compute metric
Matrix2d Metric::compute_metric(const Vector2d position,
        const std::string type) {

    Matrix2d g = Matrix2d::Zero();
    Vector2d v = position-origin;

    if (type=="polar") {
        g(0,0) = 1;
        g(1,1) = pow(sin(v[0]*M_PI),2.0);
    }
    else if (type=="a1") {
        g(0,0) = 1;
        g(1,1) = v[0];   
    }
    else if (type=="a2") {
        g(0,0) = 1;
        g(1,1) = 1./v[0];
    }

    return g;
}

void Metric::operator()( const std::vector<double> &x, 
                std::vector<double> &dxdt, 
                const double /* t */) {

    dxdt[0] = x[2]; 
    dxdt[1] = x[3]/x[0];
    dxdt[2] = 0.5*(x[3]*x[3])/(x[0]*x[0]);
    dxdt[3] = 0;
}

